### ISSUE TITLE
Tree structure suddenly destroyed - restore possible?

### ISSUE THREAD

**Issue by sbr61** [piroor/treestyletab#2305]
I had some hundred tabs open when I came across Tree Style Tab just recently. Now after I have spent lots of hours to sort my tabs into a tree structure, everything got flat within a second. (This was on an operation to move one of my tabs from a temporary to a permanent container.)

Is there a possibility to restore my previous tree structure?  (I leave Firefox open and don't create new tabs for the moment.)

## Environment

 * Platform (OS): Windows 10 Pro 1809
 * Version of Firefox: 67.0
 * Version (or revision) of Tree Style Tab: 3.0.16

**Comment by sbr61** [comment]
I was meanwhile able to restore the tree structure by copying `sessionstore-backups\previous.jsonlz4` to `sessionstore.jsonlz4` in my Firefox profile before starting Firefox again.

But then I got the same problem again, and I can describe now what happened: I had a bunch of tabs open, all with sub-pages of github.com. All of them were open in _temporary_ containers. Then I opened a new tab for github.com in my _Personal_ (permanent) container and clicked _Always open in this container_. Then I exited Firefox.

The next time I started Firefox, the tree structure was intact, the one github tab I created last was in the _Personal_ container, and all the other github tabs were still in their temporary containers. When I clicked one of them, Firefox reloaded it. This caused the Temporary Containers extension to interact and to replace the tab with a decision tab whether to open the URL in a new temporary or in the _Personal_ container. The choice (of _Personal_ in my case) caused another tab replacement. It was one of these tab replacements that caused the tree structure to be destroyed. I also noticed that it took rather long (more than 10 seconds) until the selection of my choice was accepted and the personal tab appeared, maybe because TST was in the process of checking or rebuilding the tree structure when the tab replacement by TC was made.

I solved my problem by removing the flag _Always open in this container_ for github.com in my _Personal_ container and will also report an issue for the TC extension.

**Comment by sbr61** [comment]
Piro, you did some great work by creating TST. I consider it the most important extension in the category "Productivity". But it is important that the tree structure stays intact or can easily be restored if something goes wrong. Some thoughts how to improve the situation:

1. TST: **Highest priority:** Please document on the wiki that the tree structure is part of the Firefox session and that/how it can be restored from the Firefox `sessionstore-backups`.
2. Users: Don't leave Firefox open for a too long time, e.g. exit it before hibernating your computer. Restart Firefox before/after making a big tree change or before starting a dangerous operation that could destroy the tree. This causes Firefox to create a new session backup file.
3. Users: When you have to restore a session from a backup file, copy the files from `sessionstore-backups` into a safe place outside the Firefox profile folder before you start Firefox as it will replace them on startup.
3. TST: If possible, backup the tree structure frequently and give the user the possibility to restore the tree structure or do so automatically. If not possible, cause Firefox to create session backups more frequently. If this is not possible either, describe a session backup/restore concept on the wiki (take ideas discussed here).
4. Users: Prefer to use _Create new group from tabs_ on the TST menu over using normal tabs as parents of other tabs. In case you are not able to restore the tree structure, there will be (empty) group tabs preceding their related tabs in the flat structure that will allow you to re-create the structure more easily.
5. Users: Better avoid using other extensions that manage tabs together with TST.
6. TST: As containers are a new privacy concept that interacts with tabs, users will not be able to avoid all conflicts. When testing TST, please use containers and extensions that create temporary containers or move tabs to different containers. When you experience problems, you might find a solution in TST or give users tips on the wiki how to work with such extensions.
7. TC:  I will create an issue for the _Temporary Containers_ extension about problems that TC tab replacement causes in TST.

**Comment by piroor** [comment]
Actually you can restore tabs with tree structure with sessionstore-backup, but please note that there is no guarantee. TST is now built on WebExtensions API completely, and the backend of data storage may be changed by Firefox suddenly. (For example the backend of storage API was changed from JSON file to IndexedDB.)

Wiki of this repository is public, so please add pages if you think it is required. Sadly my hands are full to maintain codes and development documents, and I seriously need help of users - peer to peer support like that.

----

I seriously hope to keep TST compatible with other tab addons, because I cannot make TST useful for everyone. Features need to be pluggable as small addons, it is very important point of the addon system of Firefox itself.

And I really need help by users themselves. Because I also a Firefox user, I cannot do such a hard work; trying many addons reported by users - but needless on my case - in my daily use. Instead, inspecting details on actual confliction by yourself will help me. If you successfully narrow down conditions to reproduce the problem you met, it should become a critical help for my debugging, even if you don't write a pull request, because I may not familiar to the addon and the usage on your case.

**Comment by piroor** [comment]
An example report is #2267 . Reports written in this format is really helpful for me.

**Comment by sbr61** [comment]
I will add something on the wiki so that users can take precautions and know how to restore the tree if something bad happens. But give me a few days, I am rather busy at the moment.

I don't know how to narrow down the problem so that it can be reproduced in a few steps with TST alone. It seems to be caused by a situation that is closely related to tab managing by _Temporary Containers_. I have no idea why TST reacts in that way.

So I think the problem can only be reproduced when TC is in use and only in a certain situation:
1. Open several tabs in temporary containers using TC all for the same domain, e.g. github.com.
2. Then open a tab in a permanent container (e.g. _Personal_) with the _Mulit-account Containers_ addon, again for github.com.
3. Then configure that permanent container to always open github.com in it (this does not change existing tabs in temporary containers with github.com).
4. Then restart Firefox.
5. Then click on one of the temporary container tabs with github.com to load it.
6. MAC will try to reload the page in the _Personal_ container, but TC will intersect and ask for confirmation whether to reload the page in the that permanent container or in a temporary container.
7. Confirm the load in the _Personal_ container.
8. One of the tab replacements happening causes TST to lose its tree structure.

I am not able to give more details than in [my second post above](https://github.com/piroor/treestyletab/issues/2305#issuecomment-500111719) and here. But I think the description will be pretty clear if you try TC yourself.

What TC actually does is creating a new (temporary) container for a single tab, either when the user manually selects from the menu to do so, or automatically on certain events and conditions, e.g. when Ctrl-clicking on a link that opens a page on a different domain. _Temporary_ means that the container (including all its data like cookies and storage) will be deleted when the tab is closed (or some minutes later so that you are able to undo if you closed the wrong tab).

Even if that functionality is useless on your case, I think using that addon would be a good stress test for TST and allow you to detect problems early.

Regarding point 4 in [my third post](https://github.com/piroor/treestyletab/issues/2305#issuecomment-500116032) on this thread, maybe this is something that a different extension should care about (frequent session backups and a restore possibility). I have not yet looked whether such an addon exists.

**Comment by piroor** [comment]
OK, I've tried to try steps you described as above.

# Steps I did

1. Start Fireflx.
2. Install TST, [Temporary Containers](https://addons.mozilla.org/firefox/addon/temporary-containers/), and [Firefox Multi-Account Containers](https://addons.mozilla.org/firefox/addon/multi-account-containers/).
3. Go to options page of Temporary Containers, and set as:
   * General => Random Container Color (ON), and click "Save".
   * Isolation => Global => Navigating in Tabs should open new Temporary Containers => Always, and click "Save".
4. Open tabs as ![image](https://user-images.githubusercontent.com/70062/59749966-f98e4880-92b8-11e9-94a5-6586012d9a0d.png):
   * A: https://github.com/piroor/treestyletab (default container)
     * A-1: https://github.com/piroor/treestyletab/issues (tmp1)
       * A-1-1: https://github.com/piroor/treestyletab/issues (tmp2)
       * A-1-2: https://github.com/piroor/treestyletab/projects (tmp3)
     * A-2: https://github.com/piroor/treestyletab/wiki (tmp4)
       * A-2-1: https://github.com/piroor/treestyletab/wiki/Code-snippets-for-custom-style-rules (tmp5)
       * A-2-2: https://github.com/piroor/treestyletab/wiki/API-for-other-addons (tmp6)
5. Open new tab B below A-2-2, from MAC's button, with the "Perosnal" container. ![image](https://user-images.githubusercontent.com/70062/59750031-22164280-92b9-11e9-9984-7f3a9d0427fb.png)
6. Activate B and load any github.com page to the tab. ![image](https://user-images.githubusercontent.com/70062/59750122-3f4b1100-92b9-11e9-88e6-5d6ef40ac930.png)
7. Open MAC's popup and check to "Always Open in Personal".
8. Exit Firefox.
9. Restart Firefox.
10. Go to the hamburger menu and choose "Restore Previous Session". ![image](https://user-images.githubusercontent.com/70062/59750323-99e46d00-92b9-11e9-9ad8-6ba55d70d8bf.png)
11. Click A-1.

# Expected result

* A-1 is restored.

# Actual result

* A, A-1-1, A-1-2, A-2, A-2-1, and A-2-2 are closed.
* MAC asks me to reopen the restored A-1 in which container. ![image](https://user-images.githubusercontent.com/70062/59750575-fa73aa00-92b9-11e9-9131-c3172b94d99e.png)
* After I choose "Open in Personal Container", A-1 is restored with the container "Personal".

**Comment by piroor** [comment]
The actual result is quite different from expected for me...

**Comment by piroor** [comment]
Second try:

# Steps (1-9 are same to the first try)

10. Tabs are restored with each container: ![image](https://user-images.githubusercontent.com/70062/59751872-41629f00-92bc-11e9-9191-6a783d4d74fa.png)
11. Click A-1.

# Expected result

* A-1 is restored at its original position (the first child of A, the parent of A-1-1 and A-1-2)

# Actual result

* The tab A-1-1 is detached from tree and moved after A-2-2, as: ![image](https://user-images.githubusercontent.com/70062/59751919-57705f80-92bc-11e9-8956-ff60cef78a51.png)

**Comment by piroor** [comment]
On the second try, even when I uncheck "Always Open in Personal" for github.com tabs, the clicked (restored) tab is detached from its original position unexpectedly. So this looks a problem of the combination: TST and Temporary Container (MAC looks not related).

**Comment by piroor** [comment]
I've introduced some changes to keep tree structure on such cases. With these changes TST keeps tree structure after tabs are reopened by Temporary Containers on the case I tested.

**Comment by sbr61** [comment]
Thank you very much. With TST version 3.1.2 I tested my original use case. Everything worked fine: all tab replacements by TC/MAC were made in the correct tree position, and the tree structure was not destroyed any more.

However, I have still seen another tree corruption with the new version, again related to TC. Fortunately, this seems to be a rare case, but this also means that it is not easy to reproduce (it may be related to timing).

These are the steps that caused the corruption in my case:
1. On the TC options page under _Isolation_/_Global_ set _Navigating in Tabs should open new Temporary Containers_ to _If the Navigation Target Domain does not exactly match the active Tabs Domain (Subdomains also get isolated)_.
2. Use MAC to open a new tab A in the _Personal_ container.
3. Open [https://www.startpage.com/](https://www.startpage.com/) in that tab.
4. Tell MAC to always open that page in the _Personal_ container.
5. Enter a search string in the Startpage input box and press Enter.
6. On the Startpage result page now displayed in tab A, change the search string in the input box above the search results and press Enter.
7. Observe that Startpage tries to open a new result page on a _different_ sub-domain of `startpage.com`, causing TC to interact.
8. Observe that TC creates a new (temporary) tab for the second result page, which is placed by TST as a child-tab A-1 of the first result page tab A.
9. Observe that MAC replaces tab A-1 with a tab A-2 in the _Personal_ container as it is configured to always open pages in the `startpage.com` domain in that container.
10. That was the point where the tree structure got corrupted

Note that the TC configuration in step 1 is not useful for Startpage, but was the reason that tab A-1 was created by TC (step 8). I am not absolutely sure about steps 9 and 10 as this happened too fast and I have not been able to reproduce the same problem again.

I have added a page on the wiki about [Recovering from a tree structure corruption](https://github.com/piroor/treestyletab/wiki/Recovering-from-a-tree-structure-corruption). I added one of your statements (see [comment above](https://github.com/piroor/treestyletab/issues/2305#issuecomment-501330477)) as a disclaimer. Please check this is OK and feel free to edit it so it matches your needs.

**Comment by irvinm** [comment]
@piroor any comments on this one?

@sbr61 nice write-up on the Wiki.  I am also wondering if we can close this item now until you have a more reproducible case on the new issue.

**Comment by sbr61** [comment]
I retested today with the newest addon versions. Step 8 [above](https://github.com/piroor/treestyletab/issues/2305#issuecomment-504705534) does not not take place anymore. Other tab replacements or additions by TC seem to be much smoother now.

My impression is that the latest changes in TC improving compatibility to TST have fixed the problem or made it less likely to occur again. So we can close the issue here.
